keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -475,24 +490,50 @@ coverage of interactions in these dimensions:
 
 Types:
 
-| OpenAPIv3 type                          | CEL type                       |
-| --------------------------------------- | ------------------------------ |
-| 'object' with Properties                | object / "message type"        |
-| 'object' with AdditionalProperties      | map                            |
-| 'array                                  | list                           |
-| 'array' with x-kubernetes-list-type=map | list with map based Equality & unique key guarantees   |
-| 'array' with x-kubernetes-list-type=set | list with set based Equality & unique entry guarantees |
-| 'boolean'                               | boolean                        |
-| 'number' (all formats)                  | double                         |
-| 'integer' (all formats)                 | int (64)                       |
-| 'null'                                  | null                           |
-| 'string'                                | string                         |
-| 'string' with format=byte               | bytes                          |
-| 'string' with format=binary             | bytes                          |
-| 'string' with format=date               | timestamp (protobuf.Timestamp) |
-| 'string' with format=datetime           | timestamp (protobuf.Timestamp) |
-
-xref: [CEL types](https://github.com/google/cel-spec/blob/master/doc/langdef.md#values), [OpenAPI types](https://swagger.io/specification/#data-types).
+| OpenAPIv3 type                                     | CEL type                                                                                                       |
+| -------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- |
+| 'object' with Properties                           | object / "message type"                                                                                        |
+| 'object' with AdditionalProperties                 | map                                                                                                            |
+| 'object' with x-kubernetes-embedded-type           | <treatment is the same as 'object' more details below>                                                         |
+| 'object' with x-kubernetes-preserve-unknown-fields | <treatment is the same as 'object', more details below>                                                        |
+| x-kubernetes-embedded-int-or-string                | object with 'intVal' (type: int) and 'strVal' (type: string) fields                                            |
```

Comment:
```suggestion
| x-kubernetes-int-or-string                | object with 'intVal' (type: int) and 'strVal' (type: string) fields                                            |
```

Reply:
can we treat this as a dynamic type that is limited to int or string values? referring the to value via a property name that doesn't match the serialized data makes expressions pretty unintuitive

Reply:
I tried this a few days ago and it works but has one caveat:

All type checking is deferred to runtime, so `intOrString == "20%" || intOrString == 10` will compile but will fail at runtime with a no "==" overload error if the intOrString is not a string.

It works fine if you guard things with a runtime type check, e.g.:

```
type(intOrString) == string && <do something as a string>
type(intOrString) == int ? <do something as an int> : <do something as a string>
```
I was a bit concerned that developers might not realize the guards are needed since there is no compile time checking, but if others are okay with it, I'll switch to the dynamic type.

Reply:
another possibility for CRDs which contain both int-or-string fields and CEL expressions is to do two passes when validating a CRD schema, once treating int-or-string as a string (and ensure expressions compile), and once treating int-or-string as an int (and ensure expressions compile)

Reply:
that keeps both readability and CRD-write-time type checking

Reply:
That's interesting.  I don't know if the type checker can handle this though.

E.g. for `type(intOrString) == int ? intOrString == 5 : intOrString == 'xyz'`, `type()` is a function that evaluates at runtime.  If it were a macro that evaluated at compile time then it would be possible for a type checker could figure it out, but I don't think that's what we have.

@TristonianJones can you confirm?

Reply:
Another thought on this.  Either way we go (a struct with two fields, or a dynamic type), we have to have a runtime guard, e.g.:

```
has(intOrString.intVal) && <subsequent expressions>
type(intOrString) == string && <subsequent expressions>
```

Both are evaluated at runtime, and without them the subsequent expressions can also fail at runtime. Either due to a unset field error or a runtime type incompatibility error.

Reply:
You've encountered homogeneous equality. The CEL language council just agreed, at long last, about how to support heterogeneous equality.  We've published one part here https://github.com/google/cel-spec/wiki/proposal-210 and are working on getting an externally friendly version of the second part ready for posting as well. 

It's a goal this quarter to introduce support for it in a flag based way, but you can expect that heterogeneous quality will just work: `intOrString == 5 || intOrString == 'xyz'` will evaluate as you would expect at runtime.

Reply:
Thanks @TristonianJones! We will keep it as a single field and then pick up language support when it is avialable.